Прохожу спринт 12

EntityVadim opened the pull request:
Ревью, прошу тебя внести комментарии по коду, может, где-то, что-то стоит поменять, улучшить и т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутые подскажешь.
Буду тебе очень сильно в этом благодарен!

EntityVadim committed:
- Добавил структуру для декодинга JSON-ответа от Unsplash — PhotoResult (а также вспомогательную структуру UrlsResult для получения полей из "urls")
- Создал структуру Photos, которая будет использоваться в UI части приложения
- Создал новый класс ImagesListService

EntityVadim committed:
- Сделал задание:
   - добавить свойство photos: [Photo] для хранения последовательности всех загруженных из сети фотографий;
   - добавить нотификацию static let didChangeNotification = Notification.Name(rawValue: "ImagesListServiceDidChange");
   - В этом классе объявить функцию fetchPhotosNextPage(), которая загружает очередную страницу; при этом

EntityVadim committed:
- Доделал сетевой слой

EntityVadim committed:
- Переместил все иконки в отдельную папку

EntityVadim committed:
- Пытаюсь реализовать ленту изображений

EntityVadim committed:
- Пытаюсь исправить проблему с лентой изображения

EntityVadim committed:
- Реализовал в проекте ImageFeed показ фотографий в соответствии с содержимым ImagesListService

EntityVadim committed:
- Работа с UI и сетевым слоем (метод changeLike проверяет, есть ли токен доступа, строит URL с использованием идентификатора фотографии, устанавливает метод HTTP в зависимости от того, должен ли лайк быть установлен или удалён, и выполняет сетевой запрос; Расширение для оповещения Notification.Name; Расширение для массива, чтобы заменить элемент Array)

EntityVadim committed:
- Добавляем взаимодействие UI с сетью

EntityVadim committed:
- Мелкие корректировки по коду

EntityVadim committed:
- Реализовал выход из профиля

EntityVadim committed:
- Мелкие корректировки по коду
- Пытаюсь добавить Scrollview

EntityVadim committed:
- Мелкие корректировки по коду
- Использовал заглушку до загрузки изображения

EntityVadim committed:
- Реализовал Scrollview

EntityVadim committed:
- Scrollview (выставил нормальное масштабирование)

EntityVadim committed:
- Выполнил пункт (по нажатию на кнопку выхода нужно показать предупреждение и запросить подтверждение выхода; если пользователь подтверждает выход из аккаунта)

EntityVadim committed:
- Исправил проблему с лайком
- Исправил проблему c didTapShareButton

EntityVadim committed:
- Уменьшил код в ширину в файлах
- Исправил проблему с отображение верной даты в ячейках
- Удалил лишние коменнтарии

EntityVadim committed:
- Мелкие исправления по коду
- Доработал обновление ячеек в таблице

EntityVadim committed:
- Сдела более четкие изображения в ленте

EntityVadim committed:
- Реализовал бесконечную ленту изображений

Nikitaaeee commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -6,27 +6,59 @@
 //
 
 import UIKit
+import Kingfisher
```
Импорты чаще всего принято располагать в алфавитном порядке.

В больших проектах могут попадаться файлы с десятками импортов (особенно, если проект разделен на множество отдельных модулей/подов), в таком случае гораздо удобнее находить/добавлять/удалять нужные импорты, если они расположены в алфавитном порядке 💡

EntityVadim replied:
Ого, это действительно очень важная информация, спасибо большое!!!

Nikitaaeee commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -6,27 +6,59 @@
-    @IBOutlet private weak var cellImage: UIImageView!
-    @IBOutlet private weak var likeButton: UIButton!
-    @IBOutlet private weak var dateLabel: UILabel!
+    @IBOutlet weak var cellImage: UIImageView!
+    @IBOutlet weak var likeButton: UIButton!
+    @IBOutlet weak var dateLabel: UILabel!
```
Первым двум свойствам можно вернуть **private** :)

Также, было бы здорово добавить **private** и для свойства `dateLabel`, а для конфигурации текста извне можно добавить новый **internal** метод для ячейки, например:
```
func updateDateLabel(withText text: String) {
   dateLabel.text = text
}
```

А затем заменить код, который конфигурирует текст лейбла:
```
if let createdAt = photo.createdAt {
   cell.updateDateLabel(withText: dateFormatter.string(from: createdAt))
} else {
   cell.updateDateLabel(withText: "Дата неизвестна")
}
```

EntityVadim replied:
Спасибо большое за совет, даже не догадывался об этом!

Nikitaaeee commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -6,27 +6,59 @@
+    
+    // MARK: - IBAction
+    
+    @IBAction func likeButtonClicked(_ sender: Any) {
+        delegate?.imageListCellDidTapLike(self)
     }
```
Тут тоже можно добавить **private**

Nikitaaeee commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,177 @@
+import UIKit
+import Kingfisher
+
+final class ImagesListService {
+    
+    static let shared = ImagesListService()
+    static let didChangeNotification = Notification.Name(rawValue: "ImagesListServiceDidChange")
+    
+    private(set) var photos: [Photo] = []
+    private var lastLoadedPage: Int?
+    private var isLoading: Bool = false
+    private var storage = OAuth2TokenStorage.shared
+    private let dateFormatter = ISO8601DateFormatter()
+    
+    private init() {}
```
Здорово, что ты не забываешь добавлять приватный инициализатор для **Синглтонов**! 😎

Nikitaaeee commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -84,21 +151,45 @@ extension ImagesListViewController: UITableViewDataSource {
+            return cell
+        }
+}
+
+extension ImagesListViewController: ImagesListCellDelegate {
+    func imageListCellDidTapLike(_ cell: ImagesListCell) {
+        guard let indexPath = tableView.indexPath(for: cell) else { return }
+        let photo = photos[indexPath.row]
+        UIBlockingProgressHUD.show()
+        imagesListService.changeLike(photoId: photo.id, isLike: !photo.isLiked) { [weak self] result in
+            guard let self = self else { return }
+            switch result {
+            case .success(_):
+                self.photos = self.imagesListService.photos
+                cell.setIsLiked(!photo.isLiked)
+            case .failure(let error):
+                print("Не удалось поставить лайк: \(error)")
+            }
+            UIBlockingProgressHUD.dismiss()
```
Тут тоже можно добавить комментарий:
`// MARK: - ImagesListCellDelegate`

Nikitaaeee reviewed (approved):
👍🏻👍🏻👍🏻

EntityVadim committed:
- Импорты расположил в алфавитном порядке;
- Добавил комментарии по коду для навигации;
- Добавил private в некоторые свойства.

EntityVadim committed:
- Прислушался к совету ревью (Также, было бы здорово добавить private и для свойства dateLabel, а для конфигурации текста извне можно добавить новый internal метод для ячейки).

EntityVadim committed:
- Поменял начальное значение прогрессбар
- Исправил проблему с индикатором загрузки (он не пропадал)
